Code/src/test/java/nl/utwente/group10/haskell/expr/ApplyTest.java

```diff
@@ -1,64 +1,95 @@
+        final Apply apply0 = new Apply(new Ident("(+)"), new Value(this.integer, "42"));
+        final Apply apply1 = new Apply(new Ident("map"), apply0);
+        final Apply apply2 = new Apply(apply1, new Value(this.integerList, "[1, 2, 3, 5, 7]"));
+
+        assertEquals("(map ((+) 42))", apply1.toHaskell());
+        assertEquals("((map ((+) 42)) [1, 2, 3, 5, 7])", apply2.toHaskell());
+
+        assertEquals(new FuncT(this.integerList, this.integerList).toHaskellType(), apply1.analyze(this.env, this.genSet).prune().toHaskellType());
+        assertEquals(this.integerList.toHaskellType(), apply2.analyze(this.env, this.genSet).prune().toHaskellType());
     }
 
     @Test
-    public void testSimpleType() throws HaskellException {
-        Func add = new EnvFunc("(+)", new FuncT(T_integer, T_integer, T_integer));
-        Apply a = new Apply(add, new Value(T_integer, "4"), new Value(T_integer, "6"));
+    public final void testZip() throws HaskellTypeError {
+        final Apply apply1 = new Apply(new Ident("zip"), new Value(this.integerList, "[1, 2, 3, 5, 7]"));
+        final Apply apply2 = new Apply(apply1, new Value(this.stringList, "[\"a\", \"b\", \"c\"]"));
+
+        assertEquals("(zip [1, 2, 3, 5, 7])", apply1.toHaskell());
+        assertEquals("((zip [1, 2, 3, 5, 7]) [\"a\", \"b\", \"c\"])", apply2.toHaskell());
 
-        assertEquals(a.toHaskell(), "(+) 4 6");
-        assertEquals(a.getType(), T_integer);
+        assertEquals(new FuncT(this.betaList, new ListT(new TupleT(this.integer, this.beta))).toHaskellType(), apply1.analyze(this.env, this.genSet).prune().toHaskellType());
+        assertEquals(new ListT(new TupleT(this.integer, this.string)).toHaskellType(), apply2.analyze(this.env, this.genSet).prune().toHaskellType());
     }
 
-//    @Test
-//    public void testCompositeType() throws HaskellException {
-//        Func massive = new EnvFunc("massive", new FuncT(T_a, T_b, T_c, new TupleT(T_a, T_b, T_c), T_string));
-//
-//        // First application
-//        Apply a1 = new Apply(massive, new Value(T_float, "1.2"), new Value(T_integer, "0"), new Value(T_string, "\"a\""));
-//        assertEquals(a1.getType().toHaskellType(), new FuncT(new TupleT(T_integer, T_float, T_string), T_string).toHaskellType());
-//        assertEquals(a1.toHaskell(), "massive 1.2 0 \"a\"");
-//
-//        // Second application
-//        Apply a2 = new Apply(a1, new Value(new TupleT(T_a, T_b, T_c), "(4.5, 3, \"b\")"));
-//        assertEquals(a2.getType(), T_string);
-//        assertEquals(a2.toHaskell(), "massive 1.2 0 \"a\" (4.5, 3, \"b\")");
-//
-//        // Third application
-//        Apply a3 = new Apply(massive, new Value(T_float, "1.2"), new Value(T_integer, "0"),
-//                new Value(T_string, "\"a\""), new Value(new TupleT(T_a, T_b, T_c), "(4.5, 3, \"b\")"));
-//        assertEquals(a3.getType(), a2.getType());
-//        assertEquals(a3.toHaskell(), a2.toHaskell());
-//    }
+    @Test(expected=HaskellTypeError.class)
```

Comment:
The last assert is there in case something goes horribly wrong - it is _probably_ never tried. The  tested aspect is also tested elsewhere.